File: OneWeekend/Python/camera.py
import numpy as np

from typedef import Vec3, Color, Hittable, Ray
from utils import random_in_uintdisk

class Camera:
    def __init__(self, lookfrom, lookat, vup, fov, ratio, aperture, focus_dist) -> None:
        theta = np.deg2rad(fov)
        h = np.tan(theta / 2)
        viewport_height = 2 * h
        viewport_width = viewport_height * ratio
        self.w = (lookfrom - lookat).unit()
        self.u = Vec3(np.cross(vup, self.w)).unit()
        self.v = Vec3(np.cross(self.w, self.u))
        self.origin = lookfrom
        self.horizontal = focus_dist * viewport_width * self.u
        self.vertical = focus_dist * viewport_height * self.v
        self.lower_left_corner = self.origin - self.horizontal / 2 - self.vertical / 2 - focus_dist * self.w
        self.lens_radius = aperture / 2

    def get_ray(self, s: float, t: float) -> Ray:
        rd = self.lens_radius * random_in_uintdisk()
        offset = self.u * rd.x() + self.v * rd.y()
        direction = self.lower_left_corner + s * self.horizontal + t * self.vertical - self.origin
        return Ray(self.origin + offset, direction - offset)

def ray_color(ray: Ray, world: Hittable, depth: int):
    record = world.hit(ray, 0.001, 1000)
    if depth < 0:
        color = Color([0, 0, 0])
    elif record is not None:
        att, ref_ray = record.material.scatter(ray, record)
        if ref_ray is not None:
            color = ray_color(ref_ray, world, depth - 1) * att
        else:
            color = Color([0, 0, 0])
    else:
        time = (ray.direction.y() + 1) / 2
        color = Color([1, 1, 1]) * (1 - time) + Color([0.5, 0.7, 1]) * time
    return color

File: OneWeekend/Rust/src/camera.rs
use std::vec;
use std::iter;

use super::vector::{Vec3, Point, Color};
use super::sence::{Sence, Hittable};
use super::utils::{self, Interval};

#[derive(Debug, Default, Clone, Copy)]
pub struct Camera {
    depth: u32,
    origin: Point,
    viewport: Viewport,
    fdist: (Vec3, Vec3)
}

impl Camera {
    pub fn new(
        origin: Point, focal: f64, fov: f64, depth: u32,
        vup: Vec3, front:Vec3, defocus: f64, w: u32, h: u32
    ) -> Self {
        let (viewport, fdist) = {
            let right = front.cross(&vup).unit();
            let up = right.cross(&front).unit();
            let center = focal * front.unit() + origin;
            let height = focal * libm::tan(utils::degrees_to_radians(fov / 2.0)) * 2.0;
            let width = height * (w as f64 / h as f64);
            let x = width * right;
            let y = height * up;
            let o = center - x / 2.0 + y / 2.0;
            let fradius = focal * libm::tan(utils::degrees_to_radians(defocus / 2.0));
            (Viewport::new(o, x, -y, w, h), (right * fradius, up * fradius))
        };
        Self { depth, origin, viewport, fdist }
    }

    pub fn render(&self, world: &Sence, width: u32, height: u32) -> Vec<Vec3> {
        let mut sample = Vec::new();
        sample.resize((width * height) as usize, Vec3::default());
        for (color, point) in iter::zip(&mut sample, &self.viewport) {
            let origin = self.defocus_disk_sample();
            let ray = Ray::new(origin, point - origin);
            *color = self.ray_color(&ray, &world, self.depth);
        }
        sample
    }

    pub fn ray_color(&self, ray: &Ray, world: &Sence, depth: u32) -> Color {
        if depth <= 0 {
            return Color::default();
        }
        if let Some(rec) = world.hit(ray, Interval::new(0.001, f64::INFINITY)) {
            if let Some((scatterd, attenuation)) = rec.material().scatter(ray, &rec) {
                return attenuation * self.ray_color(&scatterd, world, depth - 1);
            }
            return Color::default();
        }
        let unit = ray.direction().unit();
        let a = 0.5 * (unit.y() + 1.0);
        (1.0 - a) * Color::new(1.0, 1.0, 1.0) + a * Color::new(0.5, 0.7, 1.0)
    }

    fn defocus_disk_sample(&self) -> Point {
        let p = Vec3::random_in_unit_disk();
        self.origin + (p.x() * self.fdist.0) + (p.y() * self.fdist.1)
    }
}

#[derive(Debug, Default, Clone, Copy)]
struct Viewport {
    o: Point,
    x: Vec3,
    y: Vec3,
    w: u32,
    h: u32
}

impl Viewport {
    fn new(o: Point, x: Vec3, y: Vec3, w: u32, h: u32) -> Self {
        Self { o, x, y, w, h }
    }

    fn point_sample(origin: Point, i: u32, j: u32, dx: Vec3, dy: Vec3) -> Point {
        let i = i as f64 + utils::randomf(0.0, 1.0);
        let j = j as f64 + utils::randomf(0.0, 1.0);
        origin + i * dy + j * dx
    }

    fn iter(&self) -> vec::IntoIter<Vec3> {
        let dx = self.x / self.w as f64;
        let dy = self.y / self.h as f64;
        let mut points = Vec::new();
        for i in 0 .. self.h {
            for j in 0 .. self.w {
                let point = Self::point_sample(self.o, i, j, dx, dy);
                points.push(point);
            }
        }
        points.into_iter()
    }
}

impl IntoIterator for &Viewport {
    type Item = Point;
    type IntoIter = vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

#[derive(Debug, Default, Clone, Copy)]
pub struct Ray {
    origin: Point,
    direction: Vec3
}

impl Ray {
    pub fn new(origin: Point, direction: Vec3) -> Self {
        Self { origin, direction }
    }

    pub fn origin(&self) -> Point {
        self.origin
    }

    pub fn direction(&self) -> Vec3 {
        self.direction
    }

    pub fn at(&self, time: f64) -> Point {
        self.origin + time * self.direction
    }
}
